common-libs/common/src/shared/seedTools.ts

```diff
@@ -10,74 +10,96 @@ export type KeyOptions = {
   keyTypes: KeyAlgorithmType[]
 }
 
+export type ParseDecryptedSeedResult = {
+  seed: Buffer
+  didMethod: string
+  seedHexWithMethod: string
+  fullSeedHex: string
+  externalKeys?: any[]
+  metadata?: Record<string, any>
+}
+
+const ADDITIONAL_DATA_SEPARATOR = '++;additionalData:'
+const EXTERNAL_KEYS_KEY = 'keys'
+const METADATA_KEY = 'meta'
+
+/**
+ * Responsible for parsing decrypted seed
+ * As the format is changing from version to version we need to support old formats
+ * `${seedHex}++${didMethod}` - old format, backwards compatibility
+ * `${seedHex}++${didMethod}++${base64EncodedKeys, can contain '++'}` - old format, backwards compatibility
+ * `${seedHex}++${didMethod}++;additionalData:${base64EncodedData, contains both keys and metadata}` - final format
+ */
 export const getDecryptedSeedParts = (decryptedSeed: string) => {
   const seedParts = decryptedSeed.split('++')
   const seedHex = seedParts[0]
   const didMethod = seedParts[1]
-  const base64EncodedKeys = seedParts[2]
-
-  return { seedHex, didMethod, base64EncodedKeys }
+  const additionalDataSeparatorIndex = decryptedSeed.indexOf(ADDITIONAL_DATA_SEPARATOR)
+  const additionalDataIndex =
+    additionalDataSeparatorIndex > 0 && additionalDataSeparatorIndex + ADDITIONAL_DATA_SEPARATOR.length
+  const base64EncodedAdditionalData = additionalDataIndex && decryptedSeed.slice(additionalDataIndex)
+  // backwards compatibility
+  const base64EncodedKeys = !additionalDataIndex ? seedParts[2] : undefined
+
+  return { seedHex, didMethod, base64EncodedKeys, base64EncodedAdditionalData }
 }
 
-export const parseDecryptedSeed = (decryptedSeed: string) => {
-  const { seedHex, didMethod, base64EncodedKeys } = getDecryptedSeedParts(decryptedSeed)
+export const parseDecryptedSeed = (decryptedSeed: string): ParseDecryptedSeedResult => {
+  const { seedHex, didMethod, base64EncodedKeys, base64EncodedAdditionalData } = getDecryptedSeedParts(decryptedSeed)
 
   validateDidMethodSupported(didMethod)
   const seedHexWithMethod = `${seedHex}++${didMethod}`
   const seed = Buffer.from(seedHex, 'hex')
+  const additionalData = base64EncodedAdditionalData && JSON.parse(base64url.decode(base64EncodedAdditionalData))
+  const parsedFromBase64ExternalKeys = base64EncodedKeys && JSON.parse(base64url.decode(base64EncodedKeys))
+  const externalKeys = additionalData?.[EXTERNAL_KEYS_KEY] || parsedFromBase64ExternalKeys
+  const metadata = additionalData?.[METADATA_KEY]
 
-  if (!base64EncodedKeys) {
-    return {
-      seed,
-      didMethod,
-      seedHexWithMethod,
-      fullSeedHex: seedHexWithMethod,
-    }
-  }
-
-  const fullSeedHex = `${seedHexWithMethod}++${base64EncodedKeys}`
-  const externalKeys = JSON.parse(base64url.decode(base64EncodedKeys))
-
-  return { seed, didMethod, seedHexWithMethod, externalKeys, fullSeedHex }
+  return { seed, didMethod, seedHexWithMethod, externalKeys, fullSeedHex: decryptedSeed, metadata }
```

Comment:
`fullSeedHex` was used to normalize the seed when re-encrypting and re-saving it prior to this PR.

The oldest kind of seeds (that was not even `hex`, it was just a plain `seed` buffer, not even a string) is now handled outside of `seedTools`, but still we might want to normalize seeds here.

(This comment is not a call to action but just some context that can be useful to those who will look at this PR in the future)

Reply:
That this comment is not a call to action does not mean that it should be hidden.

Github allows us to merge PRs with unresolved conversations.